File: app/src/main/java/com/timotiusoktorio/pencake/ui/menu/MenuListFragmentViewModel.java
package com.timotiusoktorio.pencake.ui.menu;

import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.ViewModel;

import com.timotiusoktorio.pencake.data.model.Product;
import com.timotiusoktorio.pencake.data.model.State;
import com.timotiusoktorio.pencake.data.source.DataManager;

import java.util.List;

class MenuListFragmentViewModel extends ViewModel {

    final MutableLiveData<List<Product>> productsLiveData = new MutableLiveData<>();
    final MutableLiveData<State> stateLiveData = new MutableLiveData<>();

    private final DataManager dataManager;

    MenuListFragmentViewModel(DataManager dataManager) {
        this.dataManager = dataManager;
    }

    void loadProducts(String categoryId) {
        stateLiveData.setValue(State.LOADING);
        dataManager.fetchProducts(categoryId, new DataManager.Callback<Product>() {
            @Override
            public void onSuccess(List<Product> data) {
                productsLiveData.setValue(data);
                stateLiveData.setValue(State.SUCCESS);
            }

            @Override
            public void onError(String errorMsg) {
                stateLiveData.setValue(State.ERROR);
            }
        });
    }
}

File: app/src/main/java/com/timotiusoktorio/pencake/ui/menu/MenuListFragmentViewModel.kt
package com.timotiusoktorio.pencake.ui.menu

import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import com.timotiusoktorio.pencake.data.model.Product
import com.timotiusoktorio.pencake.data.model.State
import com.timotiusoktorio.pencake.data.source.DataManager

class MenuListFragmentViewModel(private val dataManager: DataManager, private val categoryId: String) : ViewModel() {

    val stateLiveData = MutableLiveData<State>()
    val productsLiveData = MutableLiveData<List<Product>>()

    init {
        loadProducts()
    }

    private fun loadProducts() {
        stateLiveData.value = State.LOADING
        dataManager.fetchProducts(categoryId, object : DataManager.Callback<Product> {
            override fun onSuccess(data: List<Product>) {
                productsLiveData.value = data
                stateLiveData.value = State.SUCCESS
            }

            override fun onError(errorMsg: String) {
                stateLiveData.value = State.ERROR
            }
        })
    }
}